In [12]:
import numpy as np

***
***
***

1. Вкладник поклав 50000 умовних одиниць на три різні рахунки в три різні банки. За першим рахунком виплати становитимуть 5% річних, по другому – 7% річних та по третьому 6% річних. Відомо, що через рік вкладник отримав за відсотками суму в 2250 у.о. з першого та другого банку та суму в 1400 у.о. з першого та третього банку. Скільки умовних одиниць він поклав на кожен рахунок спочатку?

x + y + z = 50 000

0.05x + 0.07y = 2 250

0.05x + 0.06z = 1 400
***
x + y + z = 50 000

5x + 7y = 225 000

5x + 6z = 140 000

In [13]:
a1 = np.array([[1, 1, 1], [5, 7, 0], [5, 0, 6]])
b1 = np.array([50, 225, 140])
x1 = np.linalg.solve(a1, b1)
x1

array([10., 25., 15.])

1й рахунок (x) = 10 000

2й рахунок (y) = 25 000

3й рахунок (z) = 15 000

***
***
***

2. На складі зберігаються мобільні телефони iPhone 6, iPhone 11, iPhone 12 моделей. Всього 1328 екземплярів. Моделей iPhone 6 на 120 менше, ніж моделей iPhone 11, та на 100 більше, ніж моделей iPhone 12. Скільки моделей кожного виду на складі?

x + y + z = 1328

y - x = 120

x - z = 100 

In [14]:
a2 = np.array([[1, 1, 1], [-1, 1, 0], [1, 0, -1]])
b2 = np.array([1328, 120, 100])
x2 = np.linalg.solve(a2, b2)
x2

array([436., 556., 336.])

iPhone 6 (x) = 436

iPhone 11 (y) = 556

iPhone 12 (z) = 336

***
***
***

3. Складіть канонічне рівняння еліпсоїда, що проходить через три точки A, B, C

A($\sqrt{3}$; 0; $\sqrt{3}$),
B($\sqrt{6}$; $\frac{1}{2}$; 0),
C(1; $\frac{1}{\sqrt{3}}$; 1)

Канонічне рівняння еліпсоїда у прямокутній декартовій системі координат має вигляд:

$\frac{x^2}{a^2}$ + $\frac{y^2}{b^2}$ + $\frac{z^2}{c^2}$ = 1

Завдання полягає у визначенні параметрів $a^{2}$, $b^{2}$ та $c^{2}$ .

$\frac{3}{a^2}$ + $\frac{3}{c^2}$ = 1

$\frac{6}{a^2}$ + $\frac{1}{{4}*b^2}$ = 1

$\frac{1}{a^2}$ + $\frac{1}{{3}*b^2}$ + $\frac{1}{c^2}$= 1
***


$\frac{3}{a^2}$ + $\frac{3}{c^2}$ = 1

$\frac{24}{a^2}$ + $\frac{1}{b^2}$ = 4

$\frac{3}{a^2}$ + $\frac{1}{b^2}$ + $\frac{3}{c^2}$ = 3
***

$\frac{1}{a^2}$ = x

$\frac{1}{b^2}$ = y

$\frac{1}{c^2}$ = z
***

3x + 3z = 1

24x + 1y = 4

3x + y + 3z = 3

In [15]:
a3 = np.array([[3, 0, 3], [24, 1, 0], [3, 1, 3]])
b3 = np.array([1, 4, 3])
x3 = np.linalg.solve(a3, b3)
x3

array([0.08333333, 2.        , 0.25      ])

$\frac{1}{a^2}$ = $\frac{1}{12}$, $\qquad$ $a^{2}$ = 12

$\frac{1}{b^2}$ = 2, $\qquad$ $b^{2}$ = $\frac{1}{2}$

$\frac{1}{c^2}$ = $\frac{1}{4}$, $\qquad$ $c^{2}$ = 4

***
***
***

4. Складіть рівняння параболи що проходить через задані три точки (1, 12), (3, 54), (-1, 2). 

Необхідно знайти коефіцієнти рівняння a, b, c

y = a $\cdot$  $x^2$ + b $\cdot$ $x$ + c

a + b + c = 12

9a + 3b + c = 54

a - b + c = 2

In [30]:
a4 = np.array([[1, 1, 1], [9, 3, 1], [1, -1, 1]])
b4 = np.array([12, 54, 2])
x4 = np.linalg.solve(a4, b4)
x4

array([4., 5., 3.])

a = 4

b = 5

c = 3

***
***
***

Складіть функцію $\color{Teal}{get\_polynom}$, яка за набором точок виду 
($x_1$, $y_1$), ($x_2$, $y_2$), ..., ($x_{n+1}$, $y_{n+1}$) буде знаходити коефіцієнти багаточлена $c_i$. Таким чином функція приймає список кортежів з координатами та повертає набір коефіцієнтів $c_1$, $c_2$, ..., $c_n$

In [31]:
from typing import List

def get_polynom(coords: List):

    n = len(coords)
    xs = [x[0] for x in coords]
    b5 = [y[1] for y in coords]
    a5 = []

    for x in xs:
        lst = []
        
        for m in range(1, n+1):
            lst.append(x ** (n-m))
        a5.append(lst)

    return np.linalg.solve(a5, b5)

get_polynom([(1, 12), (3, 54), (-1, 2)])

array([4., 5., 3.])